### ISSUE TITLE
Cannot paste password when using MSALWebviewTypeWKWebView

### ISSUE THREAD

**Issue by guidedways** [AzureAD/microsoft-authentication-library-for-objc#1501]
I don't know why this issue was closed: https://github.com/AzureAD/microsoft-authentication-library-for-objc/issues/1460

It's becoming a serious pain for all our users. Please can someone have a look into this? By simply using `MSALWebviewTypeWKWebView` you can reproduce this - CMD+V stops working in the password field and password managers don't work either.

**Comment by mipetriu** [comment]
Hi @guidedways, I'm not sure exactly why the previous issue was closed, I can double check on that.

Regarding the issue, I see in the previous issue that there were a couple of questions from our server-side auth UX team that I'm not sure were answered:

1. Is this behavior only in the Microsoft auth flow? For example, if the affected user opens a webview to Google's auth flow, are they able to paste the password there?

2. Could the affected users try the Microsoft sign-up flow as well, as opposed to the sign-in flow? Where it says "No account? Create one!", the user is taken to another form. Is pasting possible there?

I configured our MSAL test app with `MSALWebviewTypeWKWebView` as you mentioned, and unfortunately I'm not able to repro - I can paste into both the username and password fields.

**Comment by guidedways** [comment]
Please see this video. It does not work. There's no paste option in the right click menu and CMD+V does not respond. You can see me trying:

https://user-images.githubusercontent.com/202437/171852236-fb35f421-457d-42fc-93fa-ad9b6bd5645c.mp4

Here's our code:

```
                  MSALWebviewParameters *webParams = [[MSALWebviewParameters alloc] init];
                  webParams.webviewType = MSALWebviewTypeWKWebView;
                  
                  MSALInteractiveTokenParameters *parameters = [[MSALInteractiveTokenParameters alloc] initWithScopes:scopes webviewParameters:webParams];
                  
                  parameters.loginHint = loginHint;
                  parameters.promptType = MSALPromptTypePromptIfNecessary;

                  [msalAppClient acquireTokenWithParameters:parameters completionBlock:^(MSALResult * _Nullable result, NSError * _Nullable error) {
....
}];
```

**Comment by guidedways** [comment]
This does not work for any Office 365 account we've tested with. **All** our users report the same issue.

**Comment by mipetriu** [comment]
@guidedways Do you mind doing a quick test to answer the questions above? From the webview you see in the video, please choose "Sign in with another account" -> "No account? Create one!" and then double-check if pasting is possible in the account creation form. I'd like to bring this info back to our auth UX team try to understand this issue better.

**Comment by guidedways** [comment]
I think that works (without verifying) because while I was testing I entered a fake email and since that didn't exist it made me go back and enter the login again - this time I could paste! So it was surprising for me, I thought it's fixed. For the video I used my real account and indeed it failed. So yes, this is a bug when you're using an existing account.

**Comment by guidedways** [comment]
I should add that the non-webview method has been giving us a lot of headaches separately as it frequently fails and then if you try again it says "you can only have one auth session". Users have to frequently force quit their browsers to make it work. 

The web view works flawlessly except it does not allow pasting passwords or using a password manager to paste for them, so it's incredibly painful.

**Comment by mipetriu** [comment]
@guidedways I'm sorry, I just want to make sure I understand your reply:

_I entered a fake email and since that didn't exist it made me go back and enter the login again - this time I could paste!_ 
 
I'm trying to understand if different Microsoft auth UX pages (login vs. sign up) are behaving differently for you, so I can work with our auth UX team to understand if there's a pattern. Are you saying that the email field of the login page allowed pasting if your initial login hint was incorrect? What happens if you pass no login hint? Did you get a chance to try the sign-up page?

**Comment by guidedways** [comment]
Here are two videos.

1. I passed in "jerry@busyapps.onmictosoft.com" as the login hint, then in the login window, I clicked on "Sign in with a different account", tried to paste into login there, did **not** work. Clicking next and pasting into the password field also fails.

https://user-images.githubusercontent.com/202437/172037217-970c8600-381d-4611-bfc9-a03b96a5539c.mov

**Comment by guidedways** [comment]
2. Second video is when we did not pass any login hint do begin with. Even so, I cannot paste into the field. It just refuses pasting altogether. 

So basically pasting fails regardless either way, in both the login and the password fields.

https://user-images.githubusercontent.com/202437/172037231-7adbcc50-7562-48cd-b270-61cd37d7f3e4.mov

**Comment by guidedways** [comment]
Finally third video. I clicked on "Sign in with a different account" then on "Create a new account" and boom - the window just closed 😦

**Please, please** have the UX team fix the "paste" issue. These days people use password managers and keep 20 character long passwords. It's killing them.

https://user-images.githubusercontent.com/202437/172037325-b1b09da0-0430-41aa-9c49-dd0c2a2c2bce.mov

**Comment by guidedways** [comment]
If they need an app to test this with, please install our trial app (free): https://busymac.com/busycal/index.html

**Comment by mipetriu** [comment]
@guidedways Thanks for linking your trial app. I downloaded it and tested it out - I can see the same behavior you've mentioned and showed in the videos. When I configured our MSAL Test Mac app with `webviewParameters.webviewType = MSALWebviewTypeWKWebView`, prompt = login, and a passed in loginHint, I'm finding that the webview looks the same as in your app, but I can paste. That seems to suggest that a particular configuration in your webview may be causing the issue, at least compared to the vanilla webview in the MSAL Test App.

In the MSAL library, if you look at test > app > mac > `MSALAcquireTokenViewController.m`, and then find the `acquireTokenInteractive` method, I added the follow line:

```
MSALWebviewParameters *webviewParameters = [[MSALWebviewParameters alloc] initWithAuthPresentationViewController:self];

// Added here for testing
webviewParameters.webviewType = MSALWebviewTypeWKWebView;
// Added here for testing

if ([self passedInWebview])
{
    webviewParameters.customWebview = self.webView;
    webviewParameters.webviewType = MSALWebviewTypeWKWebView;
    [self.webView setHidden:NO];
}
```

Would it be possible for you to run the MSAL Test App (Mac) target or test a webview created the same way, and see if you still see the issue?

**Comment by ameyapat** [comment]
@guidedways If it is a passed in webview, using to https://developer.apple.com/documentation/uikit/uiresponder/1621105-canperformaction?language=objc to enable pasting might be an option? I found the paste option was disabled in the edit menu of the app

**Comment by guidedways** [comment]
Thanks @ameyapat I'm not sure I follow - we're not using a custom WebView. Our code is pretty straightforward. You feel this may have something to do with the FirstResponder and the edit menu simply being disabled for the popup window? However we're not controlling that window - MSAL is. Should the Edit menu handler not be present as part of MSAL's offering?

```objc
 MSALWebviewParameters *webParams = [[MSALWebviewParameters alloc] init];
                  webParams.webviewType = MSALWebviewTypeWKWebView;
                  
                  if (syncedOnce) {
                    if (@available(macOS 12.0, *)) {
                      webParams.webviewType = MSALWebviewTypeDefault;
                    }
                  }
                  
                  MSALInteractiveTokenParameters *parameters = [[MSALInteractiveTokenParameters alloc] initWithScopes:scopes webviewParameters:webParams];
                  
                  parameters.loginHint = loginHint;
                  parameters.promptType = MSALPromptTypePromptIfNecessary;
                  
                  NSString *userSuppliedEmail = exwsService.maskedLogin;
                  
                  DDLogDetailed(@"MSAL: aquiring token for: %@ (%@)", loginHint, userSuppliedEmail);
                  
                  [msalAppClient acquireTokenWithParameters:parameters completionBlock:^(MSALResult * _Nullable result, NSError * _Nullable error) { ... }
```

**Comment by guidedways** [comment]
Okay that was it 🤦‍♂️ It turns out the Paste menu was being disabled by one of the first responders given we've internally mapped CMD+V to a swizzeled method call. It was out fault after all. Thank you for bearing with me!
